### PR TITLE
Extend WKD support to include WKD URLs

### PR DESCRIPTION
**Pull request by wiktor-k**
## Description
This change allows handling Web Key Directory links by OpenKeychain (they always have scheme `https` and path starting with `/.well-known/openpgpkey/hu`).

This change also fixes two issues with WKD: 1) handling e-mails with leading/trailing spaces 2) following redirects on WKD links, just like gnupg does (some hosts like `kernel.org` use it). It also adds tests for WKD URL construction utilities.

## Motivation and Context
This change completes support of WKD in #2270 and allows importing keys over WKD using links similarly to how Facebook or HKP links are handled.

## How Has This Been Tested?

Deployed debug build on a phone and tested links mentioned in #2270.

## Screenshots (if appropriate):
After clicking on a link.
screenshot_1526978160

## Types of changes

- ✅ Bug fix (non-breaking change which fixes an issue)
- ✅ New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/util/WebKeyDirectory.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import android.support.annotation.Nullable;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class WebKeyDirectory {
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/util/WebKeyDirectory.java]: *Util

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysActivity.java [hunk 2]
```diff
@@ -208,6 +212,13 @@ protected void handleActions(@NonNull Intent intent) {
                 startListFragment(null, null, fbUsername, cloudSearchPrefs);
                 break;
             }
+            case ACTION_IMPORT_KEY_FROM_WEB_KEY_DIRECTORY: {
+                Preferences.CloudSearchPrefs cloudSearchPrefs =
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysActivity.java]: would be nicer to use a factory method here

**Comment by wiktor-k** [comment]
Okay, now it should be better. I wanted to delay AutoValue until further PR but if you're insisting I added it here. Also, I've added the preference pane, the commit looks small. I hope this entire PR did not grow too big ;-)

Thanks for your review!

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/util/Preferences.java [hunk 3]
```diff
@@ -362,62 +361,32 @@ public void setKeySignaturesTableInitialized() {
...
 
-    public static class CloudSearchPrefs implements Parcelable {
-        public final boolean searchKeyserver;
-        public final boolean searchKeybase;
-        public final boolean searchFacebook;
-        public final boolean searchWebKeyDirectory;
-        public final HkpKeyserverAddress keyserver;
+    @AutoValue
+    public static abstract class CloudSearchPrefs implements Parcelable {
+        public abstract boolean searchKeyserver();
```
Comment (wiktor-k) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/util/Preferences.java]: I didn't have better name for these accessor methods... `isSearchKeyserver`? `isSearchingKeyserver`? `shouldSearchKeyserver`? If you've got better idea tell me and I'll adjust the code.

Comment (Valodim) [reply OpenKeychain/src/main/java/org/sufficientlysecure/keychain/util/Preferences.java]: Perhaps isKeyserverEnabled?

**Comment by wiktor-k** [comment]
Okay @Valodim, this should be it. Phew...

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysActivity.java (hunk 2 above)
Comment (wiktor-k) [reply OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysActivity.java]: Ah, that's what you meant, I thought this was about `autovalue` only ;) Okay, let me fix that....

Comment (wiktor-k) [reply OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysActivity.java]: Okay fixed, there are two new methods used now: `createWebKeyDirectoryOnly` and `createKeyserverOnly`.

File: OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysActivity.java [hunk 4]
```diff
@@ -203,17 +207,23 @@ protected void handleActions(@NonNull Intent intent) {
                 String fbUsername = FacebookKeyserverClient.getUsernameFromUri(dataUri);
 
                 Preferences.CloudSearchPrefs cloudSearchPrefs =
-                        new Preferences.CloudSearchPrefs(false, true, true, false, null);
+                        Preferences.CloudSearchPrefs.create(false, true, true, false, null);
```
Comment (Valodim) [inline OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysActivity.java]: :)

**Comment by Valodim** [comment]
Nice work, thanks!

**Pull request merged**